File: src/MixNetFx/FacFunctions/Dialogs/BaseForms/frmInfoItemsManagerGeneric.vb
Imports System.Windows.Forms
Imports ActiveDev
Imports Facesso
Imports Facesso.data

Friend Class frmInfoItemsManagerGeneric(Of ItemType As IInfoItem)
    Inherits frmInfoItemsManagerBase

    Friend Delegate Sub HandleInfoItemAddDelegate()
    Friend Delegate Sub HandleInfoItemEditDelegate()
    Friend Delegate Sub HandleRefreshItemsDelegate()
    Friend Delegate Sub HandleInfoItemColumnClickDelegate()
    Friend Delegate Sub HandleInfoItemDeleteDelegate()
    Friend Delegate Sub HandleAssignCostcenterDelegate(ByVal Costcenter As CostcenterInfo)
    Friend Delegate Sub HandlePrintListDelegate()

    Private myInfoItems As InfoItems(Of ItemType)
    Private myLastColumnClickEventArgs As ColumnClickEventArgs
    Private myInfoItemAddDelegate As HandleInfoItemAddDelegate
    Private myInfoItemEditDelegate As HandleInfoItemEditDelegate
    Private myRefreshItemsDelegate As HandleRefreshItemsDelegate
    Private myInfoItemColumnClickDelegate As HandleInfoItemColumnClickDelegate
    Private myInfoItemDeleteDelegate As HandleInfoItemDeleteDelegate
    Private myAssignCostcenterDelegate As HandleAssignCostcenterDelegate
    Private myHandlePrintListDelegate As HandlePrintListDelegate

    Private myCostCenters As CostcenterInfoItems

    Sub New(ByVal InfoItemLocalizedTypeName As String)
        MyBase.new()
        Me.Text = InfoItemLocalizedTypeName + " - Stammdaten verwalten"
        For Each tsi As ToolStripItem In MyBase.EditToolStripMenuItem.DropDownItems
            If tsi.Text.Contains("%1") Then
                tsi.Text = tsi.Text.Replace("%1", InfoItemLocalizedTypeName)
            End If
        Next
        Me.Costcenters = Nothing
    End Sub

    Protected Overrides Sub OnLayout(ByVal levent As System.Windows.Forms.LayoutEventArgs)
        MyBase.OnLayout(levent)
    End Sub

    Protected Overrides Sub OnFormClosed(ByVal e As System.Windows.Forms.FormClosedEventArgs)
        MyBase.OnFormClosed(e)
    End Sub

    Friend Overrides Sub OnInfoItemDoubleClick(ByVal sender As Object, ByVal e As System.EventArgs)
        myInfoItemEditDelegate.Invoke()
        RefreshItemsManager()
    End Sub

    Friend ReadOnly Property LastColumnClickEventArgs() As ColumnClickEventArgs
        Get
            Return myLastColumnClickEventArgs
        End Get
    End Property

    Friend Property InfoItems() As InfoItems(Of ItemType)
        Get
            Return myInfoItems
        End Get
        Set(ByVal value As InfoItems(Of ItemType))
            myInfoItems = value
            If myInfoItems IsNot Nothing Then
                If myInfoItems.Count > 0 Then
                    arvInfoItems.List = value
                End If
            End If
        End Set
    End Property

    Friend Property Costcenters() As CostcenterInfoItems
        Get
            Return myCostCenters
        End Get
        Set(ByVal value As CostcenterInfoItems)
            myCostCenters = value
            If value IsNot Nothing Then
                tslCostcenters.Enabled = True
                tscCostCenters.Enabled = True
                tsbAssignCostcenter.Enabled = True
                tscCostCenters.Items.Clear()
                For Each locCcItem As CostcenterInfo In value
                    tscCostCenters.Items.Add(locCcItem)
                Next
            Else
                tscCostCenters.Items.Clear()
                tslCostcenters.Enabled = False
                tscCostCenters.Enabled = False
                tsbAssignCostcenter.Enabled = False
            End If
        End Set
    End Property

    Public ReadOnly Property SelectedInfoItem() As ItemType
        Get
            If arvInfoItems.SelectedItems.Count = 0 Then
                Return Nothing
            Else
                Return CType(arvInfoItems.SelectedItems(0).Tag, ItemType)
            End If
        End Get
    End Property

    Public ReadOnly Property SelectedInfoItems() As List(Of ItemType)
        Get
            If arvInfoItems.SelectedItems.Count = 0 Then
                Return Nothing
            Else
                Dim locSelectedItems As New List(Of ItemType)
                For Each locItem As ListViewItem In arvInfoItems.SelectedItems
                    locSelectedItems.Add(DirectCast(locItem.Tag, ItemType))
                Next
                Return locSelectedItems
            End If
        End Get
    End Property

    Friend WriteOnly Property InfoItemAddDelegate() As HandleInfoItemAddDelegate
        Set(ByVal value As HandleInfoItemAddDelegate)
            myInfoItemAddDelegate = value
            If value IsNot Nothing Then
                AddHandler ItemAddToolStripMenuItem.Click, AddressOf UICaused_ItemAdd
                AddHandler ItemAddToolStripButton.Click, AddressOf UICaused_ItemAdd
            End If
        End Set
    End Property

    Friend WriteOnly Property InfoItemDeleteDelegate() As HandleInfoItemDeleteDelegate
        Set(ByVal value As HandleInfoItemDeleteDelegate)
            myInfoItemDeleteDelegate = value
            If value IsNot Nothing Then
                AddHandler ItemDeleteToolStripMenuItem.Click, AddressOf UICaused_ItemDelete
                AddHandler ItemDeleteToolStripButton.Click, AddressOf UICaused_ItemDelete
            End If
        End Set
    End Property

    Friend WriteOnly Property InfoItemEditDelegate() As HandleInfoItemEditDelegate
        Set(ByVal value As HandleInfoItemEditDelegate)
            myInfoItemEditDelegate = value
            If value IsNot Nothing Then
                AddHandler ItemEditToolStripMenuItem.Click, AddressOf UICaused_ItemEdit
                AddHandler ItemEditToolStripButton.Click, AddressOf UICaused_ItemEdit
            End If
        End Set
    End Property

    Friend WriteOnly Property PrintListDelegate() As HandlePrintListDelegate
        Set(ByVal value As HandlePrintListDelegate)
            myHandlePrintListDelegate = value
            If value IsNot Nothing Then
                AddHandler ItemPrintToolStripButton.Click, AddressOf UICaused_PrintList
                AddHandler PrintToolStripMenuItem.Click, AddressOf UICaused_PrintList
            End If
        End Set
    End Property



    Friend WriteOnly Property RefreshItemsDelegate() As HandleRefreshItemsDelegate
        Set(ByVal value As HandleRefreshItemsDelegate)
            myRefreshItemsDelegate = value
        End Set
    End Property

    Friend WriteOnly Property InfoItemColumnClickDelegate() As HandleInfoItemColumnClickDelegate
        Set(ByVal value As HandleInfoItemColumnClickDelegate)
            myInfoItemColumnClickDelegate = value
            If value IsNot Nothing Then
                AddHandler MyBase.InfoItemsColumnClick, AddressOf UICaused_InfoItemClicked
            End If
        End Set
    End Property

    Friend WriteOnly Property AssignCostcenterDelegate() As HandleAssignCostcenterDelegate
        Set(ByVal value As HandleAssignCostcenterDelegate)
            myAssignCostcenterDelegate = value
            If value IsNot Nothing Then
                AddHandler tsbAssignCostcenter.Click, AddressOf UICaused_AssignCostcenterClicked
            End If
        End Set
    End Property

    Private Sub UICaused_ItemAdd(ByVal sender As Object, ByVal e As EventArgs)
        myInfoItemAddDelegate.Invoke()
        RefreshItemsManager()
    End Sub

    Private Sub UICaused_ItemEdit(ByVal sender As Object, ByVal e As EventArgs)
        myInfoItemEditDelegate.Invoke()
        RefreshItemsManager()
    End Sub

    Private Sub UICaused_PrintList(ByVal sender As Object, ByVal e As EventArgs)
        myHandlePrintListDelegate.Invoke()
    End Sub

    Private Sub UICaused_InfoItemClicked(ByVal sender As Object, ByVal e As System.Windows.Forms.ColumnClickEventArgs)
        myLastColumnClickEventArgs = e
        myInfoItemColumnClickDelegate.Invoke()
        RefreshItemsManager()
    End Sub

    Private Sub UICaused_ItemDelete(ByVal sender As Object, ByVal e As EventArgs)
        myInfoItemDeleteDelegate.Invoke()
        RefreshItemsManager()
    End Sub

    Private Sub UICaused_AssignCostcenterClicked(ByVal sender As Object, ByVal e As EventArgs)
        myAssignCostcenterDelegate.Invoke(DirectCast(tscCostCenters.SelectedItem, CostcenterInfo))
    End Sub

    Private Sub RefreshItemsManager()

        Dim locIItem As IInfoItem = Nothing
        If arvInfoItems.SelectedItems.Count > 0 Then
            locIItem = DirectCast(arvInfoItems.SelectedItems(0).Tag, IInfoItem)
        End If

        myRefreshItemsDelegate.Invoke()

        'Durch die Liste galoppieren, und schauen, ob wir es wiederfinden.
        'Wenn ja, dann selektieren!
        If locIItem IsNot Nothing Then
            For Each locLvi As ListViewItem In arvInfoItems.Items
                If DirectCast(locLvi.Tag, IInfoItem).DataID = locIItem.DataID Then
                    locLvi.Selected = True
                    locLvi.EnsureVisible()
                    Return
                End If
            Next
        End If
    End Sub
End Class

File: src/NetFxCSharp/Facesso.Functions/Dialogs/BaseForms/frmInfoItemsManagerGeneric.cs
using ActiveDev;
using Facesso;
using Facesso.Data;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.Functions
{
    internal class frmInfoItemsManagerGeneric<ItemType> : frmInfoItemsManagerBase where ItemType : IInfoItem
    {
        internal delegate void HandleInfoItemAddDelegate();
        internal delegate void HandleInfoItemEditDelegate();
        internal delegate void HandleRefreshItemsDelegate();
        internal delegate void HandleInfoItemColumnClickDelegate();
        internal delegate void HandleInfoItemDeleteDelegate();
        internal delegate void HandleAssignCostcenterDelegate(CostcenterInfo Costcenter);
        internal delegate void HandlePrintListDelegate();
        private InfoItems<ItemType> myInfoItems;
        private ColumnClickEventArgs myLastColumnClickEventArgs;
        private HandleInfoItemAddDelegate myInfoItemAddDelegate;
        private HandleInfoItemEditDelegate myInfoItemEditDelegate;
        private HandleRefreshItemsDelegate myRefreshItemsDelegate;
        private HandleInfoItemColumnClickDelegate myInfoItemColumnClickDelegate;
        private HandleInfoItemDeleteDelegate myInfoItemDeleteDelegate;
        private HandleAssignCostcenterDelegate myAssignCostcenterDelegate;
        private HandlePrintListDelegate myHandlePrintListDelegate;
        private CostcenterInfoItems myCostCenters;
        public frmInfoItemsManagerGeneric(string InfoItemLocalizedTypeName) : base()
        {
            this.Text = InfoItemLocalizedTypeName + " - Stammdaten verwalten";
            foreach (ToolStripItem tsi in base.EditToolStripMenuItem.DropDownItems)
            {
                if (tsi.Text.Contains("%1"))
                {
                    tsi.Text = tsi.Text.Replace("%1", InfoItemLocalizedTypeName);
                }
            }

            this.Costcenters = null;
        }

        protected override void OnLayout(System.Windows.Forms.LayoutEventArgs levent)
        {
            base.OnLayout(levent);
        }

        protected override void OnFormClosed(System.Windows.Forms.FormClosedEventArgs e)
        {
            base.OnFormClosed(e);
        }

        internal override void OnInfoItemDoubleClick(object sender, System.EventArgs e)
        {
            myInfoItemEditDelegate.Invoke();
            RefreshItemsManager();
        }

        internal ColumnClickEventArgs LastColumnClickEventArgs
        {
            get
            {
                return myLastColumnClickEventArgs;
            }
        }

        internal InfoItems<ItemType> InfoItems
        {
            get
            {
                return myInfoItems;
            }

            set
            {
                myInfoItems = value;
                if (myInfoItems != null)
                {
                    if (myInfoItems.Count > 0)
                    {
                        arvInfoItems.List = value;
                    }
                }
            }
        }

        internal CostcenterInfoItems Costcenters
        {
            get
            {
                return myCostCenters;
            }

            set
            {
                myCostCenters = value;
                if (value != null)
                {
                    tslCostcenters.Enabled = true;
                    tscCostCenters.Enabled = true;
                    tsbAssignCostcenter.Enabled = true;
                    tscCostCenters.Items.Clear();
                    foreach (CostcenterInfo locCcItem in value)
                    {
                        tscCostCenters.Items.Add(locCcItem);
                    }
                }
                else
                {
                    tscCostCenters.Items.Clear();
                    tslCostcenters.Enabled = false;
                    tscCostCenters.Enabled = false;
                    tsbAssignCostcenter.Enabled = false;
                }
            }
        }

        public ItemType SelectedInfoItem
        {
            get
            {
                if (arvInfoItems.SelectedItems.Count == 0)
                {
                    return default(ItemType);
                }
                else
                {
                    return ((ItemType)arvInfoItems.SelectedItems[0].Tag);
                }
            }
        }

        public List<ItemType> SelectedInfoItems
        {
            get
            {
                if (arvInfoItems.SelectedItems.Count == 0)
                {
                    return null;
                }
                else
                {
                    List<ItemType> locSelectedItems = new List<ItemType>();
                    foreach (ListViewItem locItem in arvInfoItems.SelectedItems)
                    {
                        locSelectedItems.Add(((ItemType)locItem.Tag));
                    }

                    return locSelectedItems;
                }
            }
        }

        internal HandleInfoItemAddDelegate InfoItemAddDelegate
        {
            set
            {
                myInfoItemAddDelegate = value;
                if (value != null)
                {
                    ItemAddToolStripMenuItem.Click += UICaused_ItemAdd;
                    ItemAddToolStripButton.Click += UICaused_ItemAdd;
                }
            }
        }

        internal HandleInfoItemDeleteDelegate InfoItemDeleteDelegate
        {
            set
            {
                myInfoItemDeleteDelegate = value;
                if (value != null)
                {
                    ItemDeleteToolStripMenuItem.Click += UICaused_ItemDelete;
                    ItemDeleteToolStripButton.Click += UICaused_ItemDelete;
                }
            }
        }

        internal HandleInfoItemEditDelegate InfoItemEditDelegate
        {
            set
            {
                myInfoItemEditDelegate = value;
                if (value != null)
                {
                    ItemEditToolStripMenuItem.Click += UICaused_ItemEdit;
                    ItemEditToolStripButton.Click += UICaused_ItemEdit;
                }
            }
        }

        internal HandlePrintListDelegate PrintListDelegate
        {
            set
            {
                myHandlePrintListDelegate = value;
                if (value != null)
                {
                    ItemPrintToolStripButton.Click += UICaused_PrintList;
                    PrintToolStripMenuItem.Click += UICaused_PrintList;
                }
            }
        }

        internal HandleRefreshItemsDelegate RefreshItemsDelegate
        {
            set
            {
                myRefreshItemsDelegate = value;
            }
        }

        internal HandleInfoItemColumnClickDelegate InfoItemColumnClickDelegate
        {
            set
            {
                myInfoItemColumnClickDelegate = value;
                if (value != null)
                {
                    base.InfoItemsColumnClick += UICaused_InfoItemClicked;
                }
            }
        }

        internal HandleAssignCostcenterDelegate AssignCostcenterDelegate
        {
            set
            {
                myAssignCostcenterDelegate = value;
                if (value != null)
                {
                    tsbAssignCostcenter.Click += UICaused_AssignCostcenterClicked;
                }
            }
        }

        private void UICaused_ItemAdd(object sender, EventArgs e)
        {
            myInfoItemAddDelegate.Invoke();
            RefreshItemsManager();
        }

        private void UICaused_ItemEdit(object sender, EventArgs e)
        {
            myInfoItemEditDelegate.Invoke();
            RefreshItemsManager();
        }

        private void UICaused_PrintList(object sender, EventArgs e)
        {
            myHandlePrintListDelegate.Invoke();
        }

        private void UICaused_InfoItemClicked(object sender, System.Windows.Forms.ColumnClickEventArgs e)
        {
            myLastColumnClickEventArgs = e;
            myInfoItemColumnClickDelegate.Invoke();
            RefreshItemsManager();
        }

        private void UICaused_ItemDelete(object sender, EventArgs e)
        {
            myInfoItemDeleteDelegate.Invoke();
            RefreshItemsManager();
        }

        private void UICaused_AssignCostcenterClicked(object sender, EventArgs e)
        {
            myAssignCostcenterDelegate.Invoke(((CostcenterInfo)tscCostCenters.SelectedItem));
        }

        private void RefreshItemsManager()
        {
            IInfoItem locIItem = null;
            if (arvInfoItems.SelectedItems.Count > 0)
            {
                locIItem = ((IInfoItem)arvInfoItems.SelectedItems[0].Tag);
            }

            myRefreshItemsDelegate.Invoke();
            //Durch die Liste galoppieren, und schauen, ob wir es wiederfinden.
            //Wenn ja, dann selektieren!
            if (locIItem != null)
            {
                foreach (ListViewItem locLvi in arvInfoItems.Items)
                {
                    if (((IInfoItem)locLvi.Tag).DataID == locIItem.DataID)
                    {
                        locLvi.Selected = true;
                        locLvi.EnsureVisible();
                        return;
                    }
                }
            }
        }
    }
}